### ISSUE TITLE
Embed full related events in Findings

### ISSUE THREAD

**Issue by mlmitch** [ocsf/ocsf-schema#995]
## Problem

Our EDR product at Cisco emits detections to customers, and the detections have the triggering endpoint activities embedded inside the detection.
We have also seen this pattern with competing EDRs.
We use this pattern to create denormalized alerts.
The underlying assertion is that alert consumers have a set of questions they immediately want answered.
"What behavior caused this alert?" is an important question to answer. 
Denormalization to answer these questions simplifies alert consumption.
It also gives better engineering outcomes since users would otherwise make more demands of the system to answer their questions.

However, it is difficult to map these detections to OCSF Findings as there is no ability to embed full Activity events within a Finding.
The schema provides `.finding_info.related_events[].uid` to point to events related to a Finding.
However, there are several difficulties with this approach.

First, not all endpoint products have unique identifiers attached to their detections or activities.
This means that the producer or mapper personas must introduce identification schemes when converting these detections to OCSF.

Second, if a producer is going to emit Findings and Activities natively in OCSF, they must provide access to separate event streams for Findings and Activities.
Additionally, the producer needs to provide guidance on how to join these two streams.

Third, and most importantly, consumers are required to implement joining Findings and Activities.
This removes the benefits of denormalization mentioned above.
Additionally, depending on the engineering complexity the consumer is capable of, we predict they simply will not recombine Findings and Activities and miss out on the security value.

## Solutions

The core requirement is to be able to embed full activity events within Findings.
We want to do this in a way that takes advantage of the OCSF type and object system.
With these requirements in mind, we've identified two different approaches that could work.

### Arrays of Events in the Finding

The first pattern is based on arrays of events.
You define an `activities` or `events` object and this object is then embedded somewhere inside the Finding.
The new object contains arrays for each activity type.
[Here is an example schema diff with the only a couple activity types to illustrate the concept](https://github.com/mlmitch/ocsf-schema/commit/2e76f1b8915eded1e7fb33532a64bea07cb66cbf).

### `just_one` Event Inside `related_event` 

The second pattern is based on creating a field for each activity type somewhere inside of the `related_event` object.
Then apply the `just_one` constraint so each related event can only have one embedded activity.
[Here is an example schema diff with the only a couple activity types to illustrate the concept](https://github.com/mlmitch/ocsf-schema/commit/ee16cc697bd81551520ad9153bc57338da41b83d)
Something nice about this implementation is that the `type` and `type_uid` fields could be set in the `related_event` to indicate which embedded activity is set.

## Additional Commentary 

The cases I've described above makes use of Activities within findings.
However, we might want to consider using events in general.
For example, I can imagine wanting to include Detection Findings inside of an Incident Finding.
This seems to be the spirit behind the `related_event` object as well.

Solutions here require being able to use OCSF events within OCSF objects, which I'm not sure is actually supported.
Some changes to the framework might be required to enable this.

The execution of the second implementation pattern somewhat depends on the interpretation of the `just_one` constraint.
I probably just missed where it is documented, but I couldn't find it.
If `just_one` means "exactly one" then different activity fields need to be encapsulated in a new object as given above so it can be optional within the `related_event`.
If `just_one` means "at most one" then the different activity fields could be put directly in related event at the risk of causing clutter.

**Comment by mlmitch** [comment]
Here is an example test detection from Cisco Secure Endpoint's behavioral detection.
It's from a dated, but simpler version of our schema.
Additionally, I've omitted much of the clutter.

The entire detection would get mapped to a Detection Finding.
Inside the detection, we have the `matched_activity.events` array.
Each entry in the array would be mapped to an OCSF Activity.
In the example, there is a HTTP network connection, so that would be mapped to an HTTP Activity.

```
{
  // <omitted>
  "data": {
    // <omitted>
    "details": {
      // <omitted>
      "detection": "apde:2020220700100",
      "name": "ExampleNetworkDetection",
      "severity": "high",
      "matched_activity": {
        // <omitted>
        "events": [
          {
            "telemetry:net_connect": {
              "app": "windbg.exe",
              "app_path": "C:\\Program Files (x86)\\Windows Kits\\10\\Debuggers\\x64",
              "http_address_list": "23.37.18.125;",
              "http_host_name": "www.cisco.com",
              "http_request_headers": "GET /site/us/en/index.html HTTP/1.1 <omitted>",
              "http_status": 200,
              "http_url": "https://www.cisco.com/site/us/en/index.html",
              "http_verb": "GET",
              "layer7_protocol": "http",
              "local_address": "10.0.0.10",
              "local_port": 59539,
              "parent_app": "explorer.exe",
              "parent_app_path": "C:\\Windows",
              "parent_pid": 20424,
              "parent_puid": 227,
              "parent_user": "user",
              "parent_user_sid": "<omitted>",
              "pid": 38584,
              "puid": 400,
              "remote_address": "23.37.18.125",
              "remote_port": 80,
              "smb_initiated": false,
              "ts": 1674599298453684000,
              "user": "user",
              "user_sid": "<omitted>"
            }
          }
        ]
      }
    }
  }
} 
```

**Comment by guy9001** [comment]
I like this. It doesn't have to be a required field, but I do want to have the **option** to embed the full event. 
For example, thinking of my customer-facing API, I'd like to give them a boolean called `include_full_related_events` that can do the JOIN for them and embed the full events, instead of them having to fetch the full events with another API call.
As another source of inspiration, we can see that Google Chronicle findings have full attached events as well - https://cloud.google.com/chronicle/docs/reference/rest/v1alpha/Finding

Regarding the implementation details, I guess we should take a similar approach to the Artifact Evidences discussion we had. I like the `just_one` suggestion.

**Comment by sunilamin** [comment]
I wanted to add some additional context from a Cisco portfolio perspective.

As part of our XDR project, we had to map many security products (CIsco and non-Cisco) to findings and in doing so found the following relationship between findings and activities (samples can be provided if necessary):

- Cisco Secure Endpoint (EDR) findings relate to `dns_activity`, `filesystem_activity`, `http_activity`, `registry_value_activity`, `registry_key_activity`, `smb_activity`, `netwrok_activity`
- Crowdstrike Falcon (EDR) finding relate to `network_activity`
- Cisco Secure Network Anlytics (NDR) finding relate to `network_activity`

- CIsco Email Threat Defence (Email) findings relate to `email_activity`

- Microsoft Defender for O365 (Email) findings relate to `email_activity`

The above is, probabaly, unsurprising and could be solved using `finding_info.related_events`. However, I feel like that assumes an implementation and is oriented from a SIEM/XDR/ASL _consumer_ perspective.

From a _source detection product_ perspective, we are looking for a convienent "on the wire" format that does not assume anything about the capabilities of the consumer. In addition, for some of these source detection products, there is no activity export outside of activites related to a dection/finding.

**Comment by mlmitch** [comment]
Thanks Sunil.

Also, here is a more complete example event with 3 triggering activities.
I've included the raw Cisco alert that customers currently receive and the same alert quickly mapped to a Detection Finding with activities embedded `related_event`.

https://gist.github.com/mlmitch/436977e52edb9ae0a1d5125cea04ac6c

Evidences are set so you can have an idea of what is there without the full activity.
Observables are omitted since the are relatively simple in OCSF.

Basically you lose the relationship between the different entities that could be in evidences.
There was talk on Tuesday of introducing some sort of relationship object in the evidences, but I think that would ultimately amount to rebuilding the activity events.

**Comment by cei1219** [comment]
Hey @mlmitch 

I am getting a little confused w/ terminology, so perhaps you can help me.

From the EDR Data Sources I have seen thus far, they all have:

- **Alert** Level
   - **Behavior** Level (multiple Behaviors in 1 Alert)
         - **Activities** Level (multiple Activities in 1 Behavior)
         
 The EDR Data Source typically has an API that provides the Alert level (but CrowdString is moving toward producing only Behaviors Level in their new API).  Crowdstrike tells us they are moving to Behavior level in order to achieve a sort of attack chain similar to what XDR is doing.

CrowdStrike might give us

- 1 Alert (this bad thing happened on an Endpoint)
    -- Behavior 1: process X did some bad stuff
            -- Activity 1: it wrote a malicious file
            -- Activity 2: it did a DNS lookup to a malicious domain
    -- Behavior 2: process Y did some bad stuff
              ....

In XDR Today, we produce one OCSF Security Finding per Behavior (and include the common Alert ID)

We populate the Finding with one Behavior (process-oriented) and any activities that process did.  The Finding can evolve over time if more Activities occur.  

I wonder how I can map your problem statement in the context of the above (what we do today).  You say:

"Our EDR product at Cisco emits detections to customers, and the detections have the triggering endpoint activities embedded inside the detection."

It sounds like you are describing only 2 levels (Alert and Behavior?  Alert and Activity?)

**Comment by mlmitch** [comment]
Good questions.

> Can you elaborate what you mean when you say "Activity"?...

I thought the term "Activity" was more standard.
Thanks for the correction. 
I was using the term because the non-Finding events I deal with have "Activity" in their name (e.g. Process Activity [1007]).
I think your definition of "non-Finding events" is correct though.

The rest is implementation dependent so what's written below is based on the idea of embedding non-Finding events in the Related Event object.

In practice, the inclusion criteria will probably be whether or not the event type has the Related Event object in it.
To avoid the indefinite embedding levels, we only want to allow embedding events that don't have the ability to have further events embedded within them.
Today, [this set of events is exactly the  "non-Finding events"](https://github.com/search?q=repo%3Aocsf%2Focsf-schema%20related_event&type=code).

> but how do you plan to implement that limitation? Explicitly specify the "allowed" event class in "Related Event", with a just_one constraint like you suggested above

I think you've got it.
In the implementation I proposed above, we define a "wrapper" object the holds `just_one` of the allowed fields in the wrapper with a field defined for each event type we allow to be embedded.

> Do you really have a way to block people from embedding raw events in Incident Findings?

I probably didn't articulate this correctly and I think we agree.
I don't want to introduce a new way of embedding Findings into Incident Findings because Incident Finding -> Finding Info Array already exists.

You're right that someone could embed non-Finding Events in Incident Finding -> Finding Info -> Related Event -> Embedded Event. Seems okay if someone really wants to do it. The number of embedding levels is still limited to 2 in this special case.

**Comment by davemcatcisco** [comment]
What we’re really talking about here is the difference between aggregation and composition, with the embedded event approach being an example of composition.

![SoWkIImgAStDuKhEIImkLb1ooynBoSnBLt0jIyqhKL98B5PmSuN9EIQNP9PbbbGYIJgwScLafgGgfEOM0CLpKtEJKekAkTnS5RHqTHMKj9H02XnIyrA0DW00](https://github.com/ocsf/ocsf-schema/assets/164061549/5b892c3d-164c-43cb-be1a-5320809840b3)

Generally speaking, a significant drawback with composition is that a parent object cannot be separated from its child objects without undergoing mutation. Not only does this have implications for runtime cost, it contravenes functional programming principles.

More specific to our problem domain is the fact that vendors frequently implement different processing streams/pipelines for different categories of ingested event, and use different persistence technologies to store and index these events. To my mind, the embedded event approach works against this architecture because it assumes or even requires that a detection finding event will be kept with its related activity events.

An approach based on the aggregation principle would utilise a schema-defined container object in which related events could be aggregated. This would satisfy the requirements of platforms that want to separate events as well as the requirements of platforms that prefer to keep the finding with the activities (i.e. they can simply keep everything in the container).

![JSvH2iCm20RWzvn2ybZe7Lg2kKVMo2PKEw8hxFRB2c5xKxzV-1VpqlmbUzZsOWOHJrA7TAWNLceHYW6cePcrijxXIbsMSTYwEP_ixpzSCJou4f0y-rr8RjGiO89fdlldcE8Klz6CSFosXRJsNXy0](https://github.com/ocsf/ocsf-schema/assets/164061549/991898ac-a6b7-433b-a2b6-524c3d641fa5)

**Comment by mlmitch** [comment]
Here is an alternative mapping of the above Cisco Secure Endpoint alert.

https://gist.github.com/mlmitch/9c4ba254a237332e020db94f6c2989c1

The mapping uses a JSON-specific event "bundle" in order to transport the Detection Finding and 3 triggering Activities together. This JSON-specific "bundle" is just putting all the events inside a JSON array. Note this isn't isn't a technique that would translate super well to something like protobuf which is why we probably want bundling guidance to be encoding specific.

Note that I introduced UUIDs at `.[0].finding_info.related_events[i].uid` and `.[i].metadata.correlation_uid` to properly link the Related Event in the finding to the Activity in the event bundle that gives the entire content of that related event.

**Comment by davemcatcisco** [comment]
> Note that I introduced UUIDs at .[0].finding_info.related_events[i].uid and .[i].metadata.correlation_uid to properly link the Related Event in the finding to the Activity in the event bundle that gives the entire content of that related event.

What's the rationale for introducing a new `metadata.correlation_uid` attribute for this purpose instead of using the `metadata.uid` attribute that is already defined in every activity event class? The description for the pre-existing attribute is:

"The logging system-assigned unique identifier of an event instance."

That sounds to me like an exact fit for our use case?

**Comment by mlmitch** [comment]
> What's the rationale for introducing a new metadata.correlation_uid attribute for this purpose instead of using the metadata.uid attribute that is already defined in every activity event class? The description for the pre-existing attribute is:

They both exist already... just a matter of choosing the correct one. Probably a good idea to clarify the intended purposes in the OCSF slack

**Comment by sunilamin** [comment]
> They both exist already... just a matter of choosing the correct one. Probably a good idea to clarify the intended purposes in the OCSF slack

I just saw this in the [documentation](https://schema.ocsf.io/1.2.0/objects/related_event?extensions=) for the Related Event object:


Unique ID | uid | Required | String | The unique identifier of the related OCSF event. This value must be equal to metadata.uid in the corresponding related event.
-- | -- | -- | -- | --

So I would say `metadata.uid` is the way to go!

**Comment by davemcatcisco** [comment]
> > They both exist already... just a matter of choosing the correct one. Probably a good idea to clarify the intended purposes in the OCSF slack
> 
> I just saw this in the [documentation](https://schema.ocsf.io/1.2.0/objects/related_event?extensions=) for the Related Event object:
> 
> Unique ID	uid	Required	String	The unique identifier of the related OCSF event. This value must be equal to metadata.uid in the corresponding related event.
> So I would say `metadata.uid` is the way to go!

That's a useful clarification introduced between 1.1 and 1.2. Thanks!

**Comment by mlmitch** [comment]
Closing as the community is going in the direction of event bundling.
See here for the addition on JSON bundling guidance.

https://github.com/ocsf/examples/pull/65
